### PR TITLE
Update gromacs to fix SIMD detection on OSX

### PR DESCRIPTION
**Pull request by mabraham**
Fix GROMACS SIMD detection on osx. /proc/cpuinfo doesn't exist, but the fallback logic was providing the SSE2 version. Now it will run other versions on suitable hardware.

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

For members of the Bioconda project, the following command is also available:

<table>
  <tr>
    <td><code>@BiocondaBot please merge</code></td>
    <td>Upload built packages/containers and merge a PR. <br />Someone must approve a PR first! <br />This reduces CI build time by reusing built artifacts.</td>
  </tr>
</table>

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

File: recipes/gromacs/2020/build.sh [hunk 1]
```diff
@@ -56,10 +56,15 @@ chmod +x "${PREFIX}/bin/${gmx}"
 { cat <<EOF
 #! /bin/bash
 
+case "$OSTYPE" in
```
Comment (mabraham) [inline recipes/gromacs/2020/build.sh:59]: Maybe I need to lift this out of the here doc

**Comment by longr** [comment]
I think you are missing the ;; after each case item.  Not strong on bash and the linter is not being helpful.

**Comment by longr** [comment]
@douglowe, you know OSX better than me, care to have a look over?

**Comment by douglowe** [comment]
We get slightly different formats of output from these two commands.

The `cat /proc/cpuinfo | grep -m1 '^flags'` returns this on linux:
```
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d
````

Whereas the `sysctl -a | grep '^hw.optional'` returns this on OSX:
```
hw.optional.floatingpoint: 1
hw.optional.mmx: 1
hw.optional.sse: 1
hw.optional.sse2: 1
hw.optional.sse3: 1
hw.optional.supplementalsse3: 1
hw.optional.sse4_1: 1
hw.optional.sse4_2: 1
hw.optional.x86_64: 1
hw.optional.aes: 1
hw.optional.avx1_0: 1
hw.optional.rdrand: 1
hw.optional.f16c: 1
hw.optional.enfstrg: 1
hw.optional.fma: 1
hw.optional.avx2_0: 1
hw.optional.bmi1: 1
hw.optional.bmi2: 1
hw.optional.rtm: 0
hw.optional.hle: 0
hw.optional.adx: 1
hw.optional.mpx: 0
hw.optional.sgx: 0
hw.optional.avx512f: 0
hw.optional.avx512cd: 0
hw.optional.avx512dq: 0
hw.optional.avx512bw: 0
hw.optional.avx512vl: 0
hw.optional.avx512ifma: 0
hw.optional.avx512vbmi: 0
```
I assume that `1` means it's an available feature, and `0` means that it is not. Perhaps we can build a grep which will match `^hw\.optional\.([a-z0-9])\: 1$`, and returns these as a single string that can then be used within the case statement in the same way as the linux string is?

**Comment by mabraham** [comment]
> We get slightly different formats of output from these two commands.
> 
> The `cat /proc/cpuinfo | grep -m1 '^flags'` returns this on linux:
> 
> ```
> flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d
> ```
> 
> Whereas the `sysctl -a | grep '^hw.optional'` returns this on OSX:
> 
> ```
> hw.optional.floatingpoint: 1
> hw.optional.mmx: 1
> hw.optional.sse: 1
> hw.optional.sse2: 1
> hw.optional.sse3: 1
> hw.optional.supplementalsse3: 1
> hw.optional.sse4_1: 1
> hw.optional.sse4_2: 1
> hw.optional.x86_64: 1
> hw.optional.aes: 1
> hw.optional.avx1_0: 1
> hw.optional.rdrand: 1
> hw.optional.f16c: 1
> hw.optional.enfstrg: 1
> hw.optional.fma: 1
> hw.optional.avx2_0: 1
> hw.optional.bmi1: 1
> hw.optional.bmi2: 1
> hw.optional.rtm: 0
> hw.optional.hle: 0
> hw.optional.adx: 1
> hw.optional.mpx: 0
> hw.optional.sgx: 0
> hw.optional.avx512f: 0
> hw.optional.avx512cd: 0
> hw.optional.avx512dq: 0
> hw.optional.avx512bw: 0
> hw.optional.avx512vl: 0
> hw.optional.avx512ifma: 0
> hw.optional.avx512vbmi: 0
> ```
> 
> I assume that `1` means it's an available feature, and `0` means that it is not. Perhaps we can build a grep which will match `^hw\.optional\.([a-z0-9])\: 1$`, and returns these as a single string that can then be used within the case statement in the same way as the linux string is?

Good point. I generalized a bit, as we have to match lines that include underscores

**Comment by longr** [comment]
If I am reading correctly, the case statement is looking for <space><flag><space> which the OSX output does not look like, it has the hw stuff before and colons after.  Perhaps instead of grep use sed?: sed -n 's/^hw.optional.\([a-z0-9_]*\)\: 1$/\1/p'

**Comment by douglowe** [comment]
Yep, I think sed might be needed here.

This is the code I ended up at - which could be made a lot more elegant!
`sysctl -a | grep '^hw.optional' | sed -E -e 's/hw\.optional\.([a-z0-9_]+): 0//g' -e 's/hw\.optional\.([a-z0-9_]+): 1/\1/g'`

It gives me these flags, however:
```
floatingpoint mmx sse sse2 sse3 supplementalsse3 sse4_1 sse4_2 x86_64 aes avx1_0 rdrand f16c enfstrg fma avx2_0 bmi1 bmi2 adx
```
Which don't match with any of the cases in the current case statement. Should we be matching with `avx2_0` in our search for `avx` (for example)?

**Comment by longr** [comment]
Okay, this gets fun. My output on Linux is:

`flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d`

  We could use pattern match in the case statements. 
  
`*\ avx512f\ *)
*\ avx2?(_0)\ *)
*\ avx?(1_0)\ *)`

**Comment by douglowe** [comment]
We could use the machdep.cpu information instead? This seems a lot more similar to the linux flag output. E.g., from `sysctl -a | grep machdep.cpu.*features` I get:
```
machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 FMA CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC MOVBE POPCNT AES PCID XSAVE OSXSAVE SEGLIM64 TSCTMR AVX1.0 RDRAND F16C
machdep.cpu.leaf7_features: RDWRFSGS TSC_THREAD_OFFSET SGX BMI1 HLE AVX2 SMEP BMI2 ERMS INVPCID RTM FPU_CSDS MPX RDSEED ADX SMAP CLFSOPT IPT MDCLEAR TSXFA IBRS STIBP L1DF SSBD
machdep.cpu.extfeatures: SYSCALL XD 1GBPAGE EM64T LAHF LZCNT PREFETCHW RDTSCP TSCI
```
Concatenating these three lists looks like it would give something similar to the linux flags.

**Comment by mabraham** [comment]
`sysctl -a | grep '^hw.optional' | sed -E -e 's/hw\.optional\.([a-z0-9_]+): 0//g' -e 's/hw\.optional\.([a-z0-9_]+): 1/\1/g'` seems rather complicated.

What problem are you solving that `sysctl -a | grep '^hw.optional\.(.*): 1$` doesn't do? It's already dropped the `: 0` lines, and because the subsequent pattern matching works fine in the x86 case we know that we don't need to clean out other stuff.

**Review by longr — COMMENTED** [review]
> `sysctl -a | grep '^hw.optional' | sed -E -e 's/hw\.optional\.([a-z0-9_]+): 0//g' -e 's/hw\.optional\.([a-z0-9_]+): 1/\1/g'` seems rather complicated.
> 
> What problem are you solving that `sysctl -a | grep '^hw.optional\.(.*): 1$` doesn't do? It's already dropped the `: 0` lines, and because the subsequent pattern matching works fine in the x86 case we know that we don't need to clean out other stuff.

My understanding was that this expression returns `hw.optional.avx1_0: 1` and case looks for `avx` surrounded by spaces, so it would not match.  We either need to change the case to be *\ avx?(1_0)\ *) or use @douglowe  suggestion to get a more linux like output.

File: recipes/gromacs/2020/build.sh [hunk 2]
```diff
@@ -56,10 +56,15 @@ chmod +x "${PREFIX}/bin/${gmx}"
...
+  darwin*) hardware_info_command="sysctl -a | grep -m1 '^hw.optional'"
+  *)       hardware_info_command="cat /proc/cpuinfo | grep -m1 '^flags'"
```
[the remark this answers is not in the document]
Comment (mabraham) [reply recipes/gromacs/2020/build.sh:59]: There is one on the next line?

**Comment by douglowe** [comment]
Yeah - my code snippet is far too complicated - I think Robin's `sed -n 's/^hw.optional.([a-z0-9_]*): 1$/\1/p'` is a much cleaner solution (if we go down that route). It is to ensure that we're passing strings to the case select which have spaces either side of our search term.

File: recipes/gromacs/2020/build.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (douglowe) [reply recipes/gromacs/2020/build.sh:59]: we need `;;` between each case statement, e.g.:
```
case "$OSTYPE" in
   darwin*) echo 'darwin!'
   ;;
   *) echo 'not darwin :('
esac
```

[the remark this answers is not in the document]
Comment (mabraham) [reply recipes/gromacs/2020/build.sh:59]: There is one. This conversation is not on the latest code version.

**Comment by mabraham** [comment]
> Yeah - my code snippet is far too complicated - I think Robin's `sed -n 's/^hw.optional.([a-z0-9_]*): 1$/\1/p'` is a much cleaner solution (if we go down that route). It is to ensure that we're passing strings to the case select which have spaces either side of our search term.

I'd like to avoid sed because Darwin's sed is from BSD and Linux's is from GNU, and they do not work the same.

**Comment by longr** [comment]
@mabraham can you confirm what avx512f is called on OSX in case we need to change the case statement for that as well?

**Comment by mabraham** [comment]
> @mabraham can you confirm what avx512f is called on OSX in case we need to change the case statement for that as well?

not sure what you mean. THis line `hw.optional.avx512f: 1` indicates support

**Comment by longr** [comment]
> > @mabraham can you confirm what avx512f is called on OSX in case we need to change the case statement for that as well?
> 
> not sure what you mean. THis line `hw.optional.avx512f: 1` indicates support

I missed that line, for some reason I copied part of it when testing and had not noticed the avx512 section.  That is fine.

**Comment by longr** [comment]
Looks like the Linux build is passing again.  I wonder if extglob is not enabled?  https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Pattern-Matching  enable with `shopt -s extglob` I believe.

**Comment by mabraham** [comment]
The nuclear option is to compile a portable binary (just like GROMACS builds and uses as part of its automated choosing of a suitable GMX_SIMD for the host machine) and use that to choose...

**Comment by mabraham** [comment]
> Looks like the Linux build is passing again. I wonder if extglob is not enabled? https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Pattern-Matching enable with `shopt -s extglob` I believe.

Good catch, but I went for simpler still!

File: recipes/gromacs/2020/build.sh [hunk 3]
```diff
@@ -115,11 +115,11 @@ function _gromacs_bin_dir() {
         "${PREFIX}/bin.AVX_512/identifyavx512fmaunits" | grep -q '2' && \
         arch='AVX_512'
     ;;
-    *avx2*)
+    *\ avx2\ | *avx_2*)
```
Comment (douglowe) [inline recipes/gromacs/2020/build.sh:118]: should the first expression be `*\ avx2\ *`? (I ask as I don't know if the search has to match the whole string, or just a subset of the string)

Comment (longr) [reply recipes/gromacs/2020/build.sh:118]: Yes, it does need a trailing `*`.

Comment (mabraham) [reply recipes/gromacs/2020/build.sh:118]: Noted, will fix once I see how the mac CI result is

**Comment by mabraham** [comment]
@BiocondaBot please add label

**Review by bgruening — APPROVED** [review]
Cool! Thanks a bunch!

**Comment by dpryan79** [comment]
@bioconda-bot please merge

**Pull request merged**